pkg/kwok/metrics/metrics.go

```diff
@@ -86,269 +79,505 @@ func (h *UpdateHandler) getNodeInfo(nodeName string) (*controllers.NodeInfo, boo
 	return nodeInfo, ok
 }
 
-func (h *UpdateHandler) init() error {
-	metric := h.metric
+func (h *UpdateHandler) getPodsInfo(nodeName string) ([]*controllers.PodInfo, bool) {
+	podsInfo, ok := h.controller.ListPods(nodeName)
+	return podsInfo, ok
+}
 
-	for _, m := range metric.Spec.Metrics {
-		m := m
+func (h *UpdateHandler) getOrRegisterGauge(metricConfig *internalversion.MetricConfig, data cel.Data) (Gauge, string, error) {
+	preparedLabels, err := h.prepareLabels(metricConfig.Labels)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to prepare labels for metric %q: %w", metricConfig.Name, err)
+	}
+
+	labels, err := h.createLabels(preparedLabels, data)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to evaluate labels: %w", err)
+	}
+
+	key := uniqueKey(metricConfig.Name, metricConfig.Kind, labels)
+	val, ok := h.gauges.Load(key)
+	if ok {
+		return val, key, nil
+	}
+
+	val = NewGauge(
+		GaugeOpts{
+			Name:        metricConfig.Name,
+			Help:        metricConfig.Help,
+			ConstLabels: labels,
+		},
+	)
+	h.gauges.Store(key, val)
+	err = h.registry.Register(val)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to register gauge %q: %w", metricConfig.Name, err)
+	}
+	return val, key, nil
+}
+
+func (h *UpdateHandler) getOrRegisterCounter(metricConfig *internalversion.MetricConfig, data cel.Data) (Counter, string, error) {
+	preparedLabels, err := h.prepareLabels(metricConfig.Labels)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to prepare labels for metric %q: %w", metricConfig.Name, err)
+	}
 
-		preparedLabels, err := h.prepareLabels(&m)
+	labels, err := h.createLabels(preparedLabels, data)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to evaluate labels: %w", err)
+	}
+
+	key := uniqueKey(metricConfig.Name, metricConfig.Kind, labels)
```

Comment:
NIT: Maybe extract this to `h.createKeyAndLabels`. It is repeated, and it really depends on the same type of data (metricConfig and data) :relaxed: